src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+                    list.ForEach(c => c._connection.Dispose());
+                    list.Clear();
+                }
+                Debug.Assert(list.Count == 0, $"Expected {nameof(list)}.{nameof(list.Count)} == 0");
+            }
+        }
 
-            public ConcurrentConnectionSet()
+        /// <summary>
+        /// Removes any unusable connections from the pool, and if the pool
+        /// is then empty and stale, disposes of it.
+        /// </summary>
+        /// <returns>
+        /// true if the pool disposes of itself; otherwise, false.
+        /// </returns>
+        public bool CleanCacheAndDisposeIfUnused()
+        {
+            List<CachedConnection> list = _idleConnections;
+            lock (SyncObj)
             {
-                _sets = new HashSet<HttpConnection>[s_numSlots];
-                for (int i = 0; i < _sets.Length; i++)
+                // Get the current time.  This is compared against each connection's last returned
+                // time to determine whether a connection is too old and should be closed.
+                DateTimeOffset now = DateTimeOffset.Now;
+
+                // Find the first item which needs to be removed.
+                int freeIndex = 0;
+                while (freeIndex < list.Count && list[freeIndex].IsUsable(now))
+                {
+                    freeIndex++;
+                }
+
+                // If freeIndex == list.Count, nothing needs to be removed.
+                // But if it's < list.Count, at least one connection needs to be purged.
+                if (freeIndex < list.Count)
                 {
-                    _sets[i] = new HashSet<HttpConnection>();
+                    // We know the connection at freeIndex is unusable, so dispose of it.
+                    list[freeIndex]._connection.Dispose();
+
+                    // Find the first item after the one to be removed that should be kept.
+                    int current = freeIndex + 1;
+                    while (current < list.Count)
+                    {
+                        // Look for the first item to be kept.  Along the way, any
+                        // that shouldn't be kept are disposed of.
+                        while (current < list.Count && !list[current].IsUsable(now))
+                        {
+                            list[current]._connection.Dispose();
+                            current++;
+                        }
+
+                        // If we found something to keep, copy it down to the known free slot.
+                        if (current < list.Count)
+                        {
+                            // copy item to the free slot
+                            list[freeIndex++] = list[current++];
+                        }
+
+                        // Keep going until there are no more good items.
+                    }
+
+                    // At this point, everything at and after freeIndex are garbage connections
```

Comment:
I'm just being nitty about wording.  I'd suggest something like "At this point, good connections have been moved below freeIndex, and garbage connections have been disposed, so clear the end of the list past freeIndex."  Or something like that.